ffi/src/handle.rs

```diff
@@ -1,235 +1,417 @@
-    unsafe fn drop_handle(handle: *mut Self) {
-        let _ = unsafe { Box::from_raw(handle) };
+    /// cbindgen:transparent-typedef
+    #[repr(transparent)]
+    pub struct Handle<H: HandleDescriptor> {
+        ptr: NonNull<H>,
+    }
+
+    unsafe impl<H: HandleDescriptor> Send for Handle<H> {}
+    unsafe impl<H: HandleDescriptor> Sync for Handle<H> {}
+
+    impl<T, M, S, H> Handle<H>
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = M, Sized = S> + HandleOps<T, M, S>,
+    {
+        /// Obtains a shared reference to this handle's underlying object.
+        pub unsafe fn as_ref(&self) -> &H::Target {
+            H::as_ref(self.ptr.cast().as_ptr())
+        }
+
+        pub unsafe fn into_inner(self) -> H::From {
+            H::into_inner(self.ptr.cast().as_ptr())
+        }
+        /// Destroys this handle.
+        pub unsafe fn drop_handle(self) {
+            drop(self.into_inner())
+        }
+    }
+
+    impl<T, S, H> From<Box<T>> for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S>
+            + HandleOps<T, True, S, From = Box<T>>,
+    {
+        fn from(val: Box<T>) -> Handle<H> {
+            let ptr = H::into_handle_ptr(val).cast();
+            Handle { ptr }
+        }
+    }
+
+    impl<T, S, H> From<Arc<T>> for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = False, Sized = S>
+            + HandleOps<T, False, S, From = Arc<T>>,
+    {
+        fn from(val: Arc<T>) -> Handle<H> {
+            let ptr = H::into_handle_ptr(val).cast();
+            Handle { ptr }
+        }
+    }
+
+    impl<T, S, H> HandleAsMut for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S> + HandleOps<T, True, S>,
+    {
+        type Target = T;
+
+        unsafe fn as_mut(&mut self) -> &mut T {
+            H::as_mut(self.ptr.cast().as_ptr())
+        }
+    }
+
+    impl<T, S, H> HandleAsArc for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = False, Sized = S> + HandleOps<T, False, S>,
+    {
+        type Target = T;
+
+        unsafe fn as_arc(&self) -> Arc<T> {
+            H::clone_arc(self.ptr.cast().as_ptr())
+        }
+    }
+
+    // NOTE: The functions in this trait assign arbitrary lifetime 'a to the references they return
+    // and are thus VERY unsafe to call directly. Instead, `Handle` methods call them, and are thus
+    // able to impose at least the semi-sane lifetime of not outliving the `Handle` itself.
+    pub trait HandleOps<T: ?Sized, M, S> {
+        type From: Sized;
+        type Raw: Sized;
+
+        fn into_handle_ptr(val: Self::From) -> NonNull<Self::Raw>;
+
+        unsafe fn as_ref<'a>(ptr: *const Self::Raw) -> &'a T;
+
+        // WARNING: unimplemented when M=False
```

Comment:
Sorry, I should have just suggested the change:

```suggestion
        // WARNING: `unimplemented!()` when M=False
```